In [7]:
path = r"C:\Users\Freun\Desktop\htcv_mgs\data\MGS_data\data"

from pathlib import Path
from PIL import Image


img_paths = list(Path(path).rglob("*.jpg"))
dims = {}
for img_path in img_paths:
    img = Image.open(img_path)
    if img.size not in dims:
        dims[img.size] = 1
    else:
        dims[img.size] += 1
    img.close()

# print("Image sizes and their counts:")
# for size, count in dims.items():
#     print(f"{size}: {count}")
    
print(f"Smalles image: {min(dims.keys())}")

print("Image sizes that are smaller than 224x224 in one or both dimensions:")
for size, count in dims.items():
    if size[0] < 224 or size[1] < 224:
        print(f"{size}: {count}")

Smalles image: (101, 101)
Image sizes that are smaller than 224x224 in one or both dimensions:
(220, 220): 2
(223, 223): 1
(209, 209): 2
(190, 190): 2
(189, 189): 1
(187, 187): 1
(197, 197): 2
(185, 185): 1
(171, 171): 1
(195, 195): 1
(101, 101): 1
(217, 217): 1
(217, 266): 1
(190, 240): 1
(269, 175): 1
(263, 205): 1
(284, 209): 1
(220, 286): 1
(244, 213): 1
(214, 194): 1
(223, 222): 1
(214, 265): 1
(180, 233): 1
(218, 267): 1
(185, 269): 1
(225, 184): 1
(214, 249): 1
(225, 203): 1
(332, 219): 1
(227, 189): 1
(213, 252): 1
(191, 241): 1
(183, 177): 1
(169, 227): 1
(219, 212): 1
(292, 220): 1
(217, 278): 1
(258, 216): 1
(255, 221): 1
(212, 192): 1
(233, 218): 1
(246, 218): 1
(215, 199): 1
(221, 237): 1
(214, 244): 1
(207, 229): 1
(270, 215): 1
(345, 222): 1
(205, 248): 1
(228, 222): 1
(311, 209): 1
(262, 189): 1
(250, 221): 1
(208, 237): 1
(222, 272): 1
(267, 206): 1
(201, 242): 1
(257, 221): 1
(260, 216): 1
(263, 216): 1
(218, 252): 1
(239, 199): 1
(274, 205): 1
(250, 205): 1
(235, 216

In [2]:
# Get normalization statistics for the image dataset 
import numpy as np
from tqdm import tqdm
from pathlib import Path
from PIL import Image

# labeled dataset
mean = np.zeros(3)
std = np.zeros(3)

path = r"C:\Users\Freun\Desktop\htcv_mgs\data\MGS_data\data"
img_paths = list(Path(path).rglob("*.jpg")) + list(Path(path).rglob("*.png"))
for img_path in tqdm(img_paths):
    img_pil = Image.open(img_path)
    img = np.array(img_pil).astype(np.float32) / 255.0
    mean += np.mean(img, axis=(0, 1))
    std += np.std(img, axis=(0, 1))
    img_pil.close()
    
print("Mean: ", mean / len(img_paths))
print("Std: ", std / len(img_paths))


# whole dataset
mean = np.zeros(3)
std = np.zeros(3)

path = r"C:\Users\Freun\Desktop\htcv_mgs\data\BMv3"
img_paths = list(Path(path).rglob("*.jpg")) + list(Path(path).rglob("*.png"))
for img_path in tqdm(img_paths):
    img_pil = Image.open(img_path)
    img = np.array(img_pil).astype(np.float32) / 255.0
    mean += np.mean(img, axis=(0, 1))
    std += np.std(img, axis=(0, 1))
    img_pil.close()

print("Mean: ", mean / len(img_paths))
print("Std: ", std / len(img_paths))

100%|██████████| 3406/3406 [01:07<00:00, 50.62it/s]  


Mean:  [0.49035715 0.43031338 0.36964953]
Std:  [0.19827825 0.18251361 0.16136015]


100%|██████████| 34939/34939 [35:59<00:00, 16.18it/s]  

Mean:  [0.48938244 0.37434759 0.25759417]
Std:  [0.23947592 0.19973618 0.14904125]


| Kanal | Mittelwert (Mean) | Standardabweichung (Std) |
|-------|-------------------|---------------------------|
| R     | 0.49035715        | 0.19827825                |
| G     | 0.43031338        | 0.18251361                |
| B     | 0.36964953        | 0.16136015                |


In [15]:
import time
import torch
import torchvision.transforms as T
from PIL import Image
import numpy as np
from pathlib import Path
from tqdm import tqdm
from skimage.metrics import structural_similarity as ssim
from skimage.metrics import peak_signal_noise_ratio as psnr

# =============== Pfad zu Bildordner ===============
image_folder = r"C:\Users\Freun\Desktop\htcv_mgs\data\MGS_data\data"
image_paths = list(Path(image_folder).rglob("*.jpg"))

# =============== Transformationen vorbereiten ===============
to_tensor = T.ToTensor()
resize_224 = {
    "bicubic": T.Resize((224, 224), interpolation=Image.BICUBIC),
    "bilinear": T.Resize((224, 224), interpolation=Image.BILINEAR),
    "nearest": T.Resize((224, 224), interpolation=Image.NEAREST)
}

# =============== Methoden ===============
def method_bicubic(img): return resize_224["bicubic"](img)
def method_bilinear(img): return resize_224["bilinear"](img)
def method_nearest(img): return resize_224["nearest"](img)

def method_stride(img_tensor):
    cropped = center_crop_to_square_multiple(img_tensor, multiple=224)
    if cropped is None:
        return None
    factor = cropped.shape[1] // 224
    return cropped[:, ::factor, ::factor]

def center_crop_to_square_multiple(img_tensor, multiple=224):
    _, h, w = img_tensor.shape
    min_dim = min(h, w)
    target_size = (min_dim // multiple) * multiple
    if target_size < multiple:
        return None
    top = (h - target_size) // 2
    left = (w - target_size) // 2
    return img_tensor[:, top:top+target_size, left:left+target_size]

def to_numpy(img):
    if isinstance(img, Image.Image):
        img = np.array(img)
    elif isinstance(img, torch.Tensor):
        img = img.detach().cpu()
        if img.max() <= 1.0:
            img = img * 255
        img = img.permute(1, 2, 0).numpy()
    return np.clip(img, 0, 255).astype(np.uint8)

def compare_metrics(ref, other):
    return psnr(ref, other), ssim(ref, other, channel_axis=-1)

# =============== Metrik-Tracking ===============
times = {"bicubic": [], "bilinear": [], "stride": [], "nearest": []}
psnr_list = {"bilinear": [], "stride": [], "nearest": []}
ssim_list = {"bilinear": [], "stride": [], "nearest": []}

# =============== Verarbeitungsschleife ===============
print(f"Starte Verarbeitung von {len(image_paths)} Bildern...\n")
for path in tqdm(image_paths):
    try:
        img = Image.open(path).convert("RGB")
        img_tensor = to_tensor(img)

        # Methode 1 – Bicubic (Referenz)
        t0 = time.time()
        ref_out = method_bicubic(img)
        t1 = time.time()
        times["bicubic"].append((t1 - t0) * 1000)
        ref_np = to_numpy(ref_out)

        # Methode 2 – Bilinear
        t0 = time.time()
        out_bil = method_bilinear(img)
        t1 = time.time()
        times["bilinear"].append((t1 - t0) * 1000)
        psnr_val, ssim_val = compare_metrics(ref_np, to_numpy(out_bil))
        psnr_list["bilinear"].append(psnr_val)
        ssim_list["bilinear"].append(ssim_val)

        # Methode 3 – Stride
        t0 = time.time()
        out_stride = method_stride(img_tensor)
        t1 = time.time()
        if out_stride is None:
            print(f"Bild {path.name} zu klein für Stride.")
            continue
        times["stride"].append((t1 - t0) * 1000)
        psnr_val, ssim_val = compare_metrics(ref_np, to_numpy(out_stride))
        psnr_list["stride"].append(psnr_val)
        ssim_list["stride"].append(ssim_val)

        # Methode 4 – Nearest
        t0 = time.time()
        out_near = method_nearest(img)
        t1 = time.time()
        times["nearest"].append((t1 - t0) * 1000)
        psnr_val, ssim_val = compare_metrics(ref_np, to_numpy(out_near))
        psnr_list["nearest"].append(psnr_val)
        ssim_list["nearest"].append(ssim_val)

    except Exception as e:
        print(f"Fehler bei {path}: {e}")

# =============== Ergebnisse ausgeben ===============
def summarize(name, values):
    values = np.array(values)
    return f"{name:<15}: Mittelwert = {values.mean():.4f}, Std = {values.std():.4f}"

print("\n--- Benchmark-Ergebnisse ---")
for key in times:
    print(summarize(f"Zeit {key}", times[key]))

print()
for key in psnr_list:
    print(summarize(f"PSNR {key}", psnr_list[key]))
    print(summarize(f"SSIM {key}", ssim_list[key]))


Starte Verarbeitung von 3406 Bildern...



 39%|███▉      | 1334/3406 [00:40<00:41, 49.97it/s]

Bild aw_20201009_130446211_mouse_l_06576.jpg zu klein für Stride.


 41%|████      | 1393/3406 [00:42<00:39, 50.39it/s]

Bild aw_20201027_091716929_mouse_l_03719.jpg zu klein für Stride.


 43%|████▎     | 1462/3406 [00:43<00:37, 51.44it/s]

Bild aw_20201110_092907356_mouse_r_04589.jpg zu klein für Stride.


 44%|████▍     | 1497/3406 [00:44<00:37, 50.64it/s]

Bild aw_20201124_095320521_mouse_r_00800.jpg zu klein für Stride.
Bild aw_20201125_102052423_mouse_r_02808.jpg zu klein für Stride.


 51%|█████▏    | 1749/3406 [00:49<00:32, 50.74it/s]

Bild jw_20210504_133136967_00260.jpg zu klein für Stride.


 54%|█████▎    | 1828/3406 [00:50<00:31, 50.69it/s]

Bild jw_20210507_092114543_10639.jpg zu klein für Stride.


 60%|█████▉    | 2033/3406 [00:55<00:27, 50.13it/s]

Bild jw_20211104_103423371_00772.jpg zu klein für Stride.


 60%|██████    | 2051/3406 [00:55<00:26, 50.47it/s]

Bild jw_20211107_085213746_00036.jpg zu klein für Stride.


 61%|██████    | 2069/3406 [00:55<00:26, 50.76it/s]

Bild jw_20211107_085213746_00941.jpg zu klein für Stride.


 61%|██████▏   | 2087/3406 [00:56<00:25, 52.73it/s]

Bild jw_20211107_085213746_02702.jpg zu klein für Stride.
Bild jw_20211107_085213746_03420.jpg zu klein für Stride.
Bild jw_20211107_085213746_04047.jpg zu klein für Stride.


 62%|██████▏   | 2100/3406 [00:56<00:24, 53.35it/s]

Bild jw_20211107_085213746_05472.jpg zu klein für Stride.


 63%|██████▎   | 2131/3406 [00:57<00:23, 53.33it/s]

Bild jw_20211110_071115192_03209.jpg zu klein für Stride.
Bild jw_20211110_071115192_03301.jpg zu klein für Stride.


 65%|██████▍   | 2199/3406 [00:58<00:24, 49.66it/s]

Bild lw_1-9c.JPG zu klein für Stride.


 65%|██████▍   | 2210/3406 [00:58<00:23, 51.33it/s]

Bild lw_11-14a.JPG zu klein für Stride.


 67%|██████▋   | 2279/3406 [01:00<00:24, 46.90it/s]

Bild lw_12-2a.JPG zu klein für Stride.


 68%|██████▊   | 2302/3406 [01:00<00:20, 53.74it/s]

Bild lw_13-11b.jpg zu klein für Stride.
Bild lw_13-11c.jpg zu klein für Stride.
Bild lw_13-12a.jpg zu klein für Stride.
Bild lw_13-13b.JPG zu klein für Stride.


 68%|██████▊   | 2321/3406 [01:00<00:19, 55.22it/s]

Bild lw_13-17c.jpg zu klein für Stride.
Bild lw_13-18b.jpg zu klein für Stride.
Bild lw_13-1c.jpg zu klein für Stride.


 68%|██████▊   | 2327/3406 [01:01<00:20, 52.02it/s]

Bild lw_13-20c.jpg zu klein für Stride.


 69%|██████▊   | 2339/3406 [01:01<00:20, 52.14it/s]

Bild lw_13-5a.jpg zu klein für Stride.
Bild lw_13-7b.jpg zu klein für Stride.


 69%|██████▉   | 2352/3406 [01:01<00:19, 54.67it/s]

Bild lw_13-9a.jpg zu klein für Stride.
Bild lw_16-10c.JPG zu klein für Stride.
Bild lw_16-11b.JPG zu klein für Stride.
Bild lw_16-11c.JPG zu klein für Stride.
Bild lw_16-12b.JPG zu klein für Stride.
Bild lw_16-13a.JPG zu klein für Stride.


 69%|██████▉   | 2367/3406 [01:01<00:17, 59.09it/s]

Bild lw_16-14a.JPG zu klein für Stride.
Bild lw_16-15a.JPG zu klein für Stride.
Bild lw_16-15b.JPG zu klein für Stride.
Bild lw_16-15c.JPG zu klein für Stride.
Bild lw_16-16c.JPG zu klein für Stride.
Bild lw_16-17b.JPG zu klein für Stride.


 70%|██████▉   | 2381/3406 [01:01<00:16, 61.33it/s]

Bild lw_16-18b.JPG zu klein für Stride.
Bild lw_16-1a.JPG zu klein für Stride.
Bild lw_16-1b.JPG zu klein für Stride.
Bild lw_16-1c.JPG zu klein für Stride.
Bild lw_16-20c.JPG zu klein für Stride.


 71%|███████   | 2402/3406 [01:02<00:16, 61.01it/s]

Bild lw_16-3c.JPG zu klein für Stride.
Bild lw_16-5a.JPG zu klein für Stride.
Bild lw_16-5b.JPG zu klein für Stride.
Bild lw_16-7a.JPG zu klein für Stride.
Bild lw_16-7c.JPG zu klein für Stride.


 71%|███████   | 2415/3406 [01:02<00:17, 55.11it/s]

Bild lw_16-8c.JPG zu klein für Stride.


 72%|███████▏  | 2446/3406 [01:03<00:18, 52.51it/s]

Bild lw_19-19b.JPG zu klein für Stride.
Bild lw_19-19c.JPG zu klein für Stride.


 73%|███████▎  | 2478/3406 [01:03<00:16, 55.54it/s]

Bild lw_2-11a.JPG zu klein für Stride.
Bild lw_2-11b.JPG zu klein für Stride.
Bild lw_2-14a.JPG zu klein für Stride.
Bild lw_2-14c.JPG zu klein für Stride.


 73%|███████▎  | 2502/3406 [01:04<00:17, 50.44it/s]

Bild lw_2-3b.JPG zu klein für Stride.


 74%|███████▍  | 2522/3406 [01:04<00:15, 57.14it/s]

Bild lw_3-10b.JPG zu klein für Stride.
Bild lw_3-12a.JPG zu klein für Stride.
Bild lw_3-13a.JPG zu klein für Stride.
Bild lw_3-13c.JPG zu klein für Stride.
Bild lw_3-14c.JPG zu klein für Stride.


 74%|███████▍  | 2534/3406 [01:04<00:15, 55.82it/s]

Bild lw_3-16c.JPG zu klein für Stride.
Bild lw_3-17c.JPG zu klein für Stride.
Bild lw_3-18a.JPG zu klein für Stride.
Bild lw_3-1a.JPG zu klein für Stride.


 75%|███████▍  | 2552/3406 [01:05<00:15, 54.31it/s]

Bild lw_3-2b.JPG zu klein für Stride.


 75%|███████▌  | 2564/3406 [01:05<00:15, 53.23it/s]

Bild lw_3-6b.JPG zu klein für Stride.


100%|██████████| 3406/3406 [01:30<00:00, 37.53it/s]


--- Benchmark-Ergebnisse ---
Zeit bicubic   : Mittelwert = 2.7382, Std = 2.3283
Zeit bilinear  : Mittelwert = 1.3936, Std = 1.2276
Zeit stride    : Mittelwert = 0.0558, Std = 0.2297
Zeit nearest   : Mittelwert = 0.1169, Std = 0.3198

PSNR bilinear  : Mittelwert = 50.8343, Std = 2.2208
SSIM bilinear  : Mittelwert = 0.9966, Std = 0.0008
PSNR stride    : Mittelwert = 19.9579, Std = 5.8874
SSIM stride    : Mittelwert = 0.5692, Std = 0.1375
PSNR nearest   : Mittelwert = 40.9852, Std = 2.3799
SSIM nearest   : Mittelwert = 0.9675, Std = 0.0157


### Benchmark-Ergebnisse
Bicubic resized images were taken as benchmark results and the resized versions of the other interpolation techniques are compared via SSIM and PSNR with the bicubic result

| Methode   | Zeit (Mittelwert) | Zeit (Std) | PSNR (Mittelwert) | PSNR (Std) | SSIM (Mittelwert) | SSIM (Std) |
|-----------|-------------------|------------|-------------------|------------|-------------------|------------|
| bicubic   | 2.7382            | 2.3283     | -                 | –          | –                 | –          |   
| bilinear  | 1.3936            | 1.2276     | 50.8343           | 2.2208     | 0.9966            | 0.0008     |
| stride    | 0.0558            | 0.2297     | 19.9579           | 5.8874     | 0.5692            | 0.1375     |
| nearest   | 0.1169            | 0.3198     | 40.9852           | 2.3799     | 0.9675            | 0.0157     |


- Stride is much faster than the other two but results large differences in terms of pixel values (psnr) and also large losses in terms of structural similarity (meaning the visible noise and loss of structural informations that a human would also be able to see)
- Nearest Interpolation results in nearly the same loss then Bilinear Interpolation meaning that the images tend to have few texture, are mainly centered and have no hard lines or edge which the nearest interpolation would distort. Therefor for our image dataset the nearest interpolation technique seems the by the best traid of in terms of speed and interpolation accuracy. The results of the model have to tested after the hyperparameters are found.